keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+      * [Namespace Match](#namespace-match)
+      * [Apply the template](#apply-the-template)
+      * [Schema validation](#schema-validation)
+      * [Opt-out](#opt-out)
+      * [Self-serviced namespace creation](#self-serviced-namespace-creation)
+      * [Readiness of a namespace](#readiness-of-a-namespace)
+    * [Security Consideration](#security-consideration)
+      * [Opt-out is privileged](#opt-out-is-privileged)
+      * [Race Condition](#race-condition)
+      * [Manual Update of objects](#manual-update-of-objects)
+* [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+Namespace Population is an automated mechanism to make sure the predefined policy objects
+(e.g. NetworkPolicy, Role, RoleBinding) are present in selected namespaces.
+
+## Motivation
+
+Kubernetes users create namespaces (self-service namespace creation) and want to
+ensure security/isolation policies are enforced between them and other users.
+Just like addon manager creates cluster-scope policy objects like PodSecurityPolicy,
+ClusterRole, ClusterRoleBinding at cluster creation time.
+It creates namespace-scope policy objects during namespace creation time.
+
+### Goals
+
+The namespace population mechanism described in this proposal is effective to
+populate identical configurations into large number of namespaces.
+It's also effective to populate different sets of predefined configurations into
+different sets of namespaces.
+The mechanism is agnostic to object types as long as they are namespace scoped.
+The mechanism works in a single cluster.
+
+### Non-Goals
+
+The namespace population mechanism is NOT effective for customizing namespaces
+individually.
+No dynamic objects (generating the object definition during population,
+e.g. string substitution) is supported - all objects are defined statically, 
+except `$(CREATOR)` and `$(NAMESPACE)`.
+No validation is performed over static object definitions, until they are applied.
+The mechanism does NOT work across clusters.
+
+## Proposal
+
+The namespace population is performed by defining one or more cluster-scope
+Custom Resource with kind `NamespaceTemplate`,
+and deploy a controller in the cluster.
+
+The Custom Resource Definition of `NamespaceTemplate`:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: namespacetemplates.policy
+spec:
+  group: policy
+  version: v1alpha1
+  scope: Cluster
+  names:
+    plural: namespacetemplates
+    singular: namespacetemplate
+    kind: NamespaceTemplate
+    shortNames:
+    - nstpl
+```
+
+The example of `NamespaceTemplate`:
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: NamespaceTemplate
+metadata:
+  name: default
+spec:
+  namespaces:
+    # labelSelector selects namespaces by labels to populate defined objects
+    labelSelector: {}
+    # excludes lists names of namespaces which must be excluded from being applied 
+    # with current namespace template.
+    excludes: []
+  common:
```

Comment:
I'd rather not do this - it's too clever.  Users can easily duplicate labels and be precise.

Reply:
The intention was for other components to specify labels they are interested in when integrating namespace population mechanism, not for users. I will remove it, as it's not necessary for now.